### PR TITLE
feat(ui): add ability to create volume groups

### PR DESCRIPTION
**Pull request by Caleb-Ellis**
## Done

- Added `BulkActions` component + types to house actions that can be performed on multiple storage devices at once
- Added `CreateVolumeGroup` component which is a form for creating a volume group from selected eligible devices

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to the storage tab of a Ready or Allocated machine
- Change the storage layout to blank
- Without any checkboxes checked, hover over the disabled "Create volume group" button and see there is a tooltip
- Check the disk checkbox and then click the enabled "Create volume group" button and submit
- A new volume group should appear in Available, and the disk should have moved to Used (alongside an automatically created partition)
- Remove the volume group from the actions dropdown so the disk moves back to Available
- Create two disk partitions from the actions dropdown
- Check the checkbox for the disk and see that the "Create volume group" button stays disabled
- Uncheck the disk checkbox and check the two partition checkboxes
- The "Create volume group" button should be enabled
- Click it and see that the table shows the details of the two partitions and calculates the total size
- Submit the form and check that a volume group shows up in Available and the two partitions move to Used

## Fixes

Fixes canonical-web-and-design/MAAS-squad#2167

## Screenshot

Screenshot_2020-12-11 fun-mammal maas storage bolla MAAS

### CODE REVIEW

File: ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/BulkActions/BulkActions.tsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+          data-test="create-volume-group"
+          message={
+            !createVgEnabled
+              ? "Select one or more unpartitioned and unformatted storage devices to create a volume group."
+              : null
+          }
+          position="top-left"
+        >
+          <Button
+            appearance="neutral"
```
Comment (anthonydillon) [inline ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/BulkActions/BulkActions.tsx:47]: Should this be positive?

**Review by anthonydillon — APPROVED** [review]
LGTM buy a few small comments to address

File: ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/BulkActions/BulkActions.tsx (hunk 1 above)
Comment (Caleb-Ellis) [reply ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/BulkActions/BulkActions.tsx:47]: I don't think so, it'll be alongside a few other buttons with equal importance and I don't think they'd all be positive. Besides I'm just reimplementing how it is in the angular version.

File: ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/AvailableStorageTable.tsx [hunk 2]
```diff
@@ -52,6 +58,17 @@ type Props = {
...
+/**
+ * Returns whether a storage device is available.
+ * @param storageDevice - the disk or partition to check.
+ * @returns whether a storage device is available.
+ */
+const isAvailable = (storageDevice: Disk | Partition) =>
+  !isDatastore(storageDevice.filesystem) &&
+  (isDisk(storageDevice)
+    ? diskAvailable(storageDevice)
+    : partitionAvailable(storageDevice));
```
[the remark this answers is not in the document]
Comment (Caleb-Ellis) [reply ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/AvailableStorageTable.tsx:70]: Good call! Got too focused on keeping it short

File: ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/BulkActions/CreateVolumeGroup/CreateVolumeGroup.tsx [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    systemId,
+    "creatingVolumeGroup",
+    "createVolumeGroup",
+    () => closeForm()
+  );
+  const totalSize = selected.reduce((sum, device) => (sum += device.size), 0);
+
+  if (machine && "disks" in machine) {
+    const initialName = `vg${machine.disks.reduce(
+      (sum, d) => (d.type === DiskTypes.VOLUME_GROUP ? sum + 1 : sum),
```
[the remark this answers is not in the document]
Comment (Caleb-Ellis) [reply ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/BulkActions/CreateVolumeGroup/CreateVolumeGroup.tsx:58]: Good idea!

File: ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/BulkActions/CreateVolumeGroup/CreateVolumeGroup.tsx [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+    return (
+      <FormCard sidebar={false}>
+        <FormikForm
+          allowUnchanged
+          buttons={FormCardButtons}
+          cleanup={machineActions.cleanup}
+          errors={errors}
+          initialValues={{
+            name: initialName,
```
[the remark this answers is not in the document]
Comment (Caleb-Ellis) [reply ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/BulkActions/CreateVolumeGroup/CreateVolumeGroup.tsx:70]: I'll fix this with the name generator function so it always returns something

**Pull request merged**